File: app/src/Shared/Application/Bus/CommandBus.php
<?php

declare(strict_types=1);

namespace Src\Shared\Application\Bus;

/**
 * COMMAND BUS (CQRS Pattern)
 * ==========================
 * Command Bus — Command-ları müvafiq Handler-ə yönləndirən vasitədir.
 *
 * CQRS-də ROLU:
 * - CQRS yazma (Command) və oxuma (Query) əməliyyatlarını ayırır.
 * - Command Bus yalnız YAZMA əməliyyatları üçündür.
 * - CreateOrderCommand → CommandBus → CreateOrderHandler
 *
 * MIDDLEWARE PIPELINE:
 * Command Bus-a middleware əlavə etmək olur:
 * Command → [Validation] → [Logging] → [Transaction] → Handler
 * Hər middleware Command-ı emal edib növbəti middleware-ə ötürür.
 */
interface CommandBus
{
    /**
     * Command-ı icra et.
     * Bus command-ı middleware pipeline-dan keçirib Handler-ə çatdırır.
     *
     * @throws \Exception Handler və ya middleware xəta atarsa
     */
    public function dispatch(Command $command): mixed;
}

File: app/spring/src/main/java/az/ecommerce/shared/application/bus/CommandBus.java
package az.ecommerce.shared.application.bus;

/**
 * Laravel: src/Shared/Application/Bus/CommandBus.php (interface)
 * Spring: SimpleCommandBus impl bunu @Service kimi qeyd edir.
 *
 * Controller-də: commandBus.dispatch(new CreateOrderCommand(...));
 */
public interface CommandBus {

    <R> R dispatch(Command<R> command);
}
